If the input is contiguous, short-circuit infer_size_dv in reshape

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95216

The main improvement is that this avoids guards from infer_size_dv,
although this also counts as a minor perf improvement too.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
If the input is contiguous, short-circuit infer_size_dv in reshape

The main improvement is that this avoids guards from infer_size_dv,
although this also counts as a minor perf improvement too.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD reviewed (approved):
I guess that's ok as long as you're not doing anything funky with `dispatch_sizes_strides_policy`.

You do need to add the same `    if (!self.is_xla() && !self.is_lazy() && !self.is_ipu() && !at::isTensorSubclassLike(self)) {` as below though.

ezyang commented:
I think it is sound to omit the checks. Let us enumerate the cases:

* What about mkldnn? MKLDNN tensors are never contiguous, thus they will never hit this case.
* What about xla/lazy/ipu/tensor subclass short circuit? In fact, the whole point of the condition there is to ensure we hit a proper view. When we short circuit is contiguous, we always go to view. So we apply this behavior in that case
* What about the non short circuit case? It's always safe to go the slower route. In fact, it is required, because we didn't actually compute the stride

ezyang commented:
> What about mkldnn? MKLDNN tensors are never contiguous, thus they will never hit this case.

lol but CI HAS PROVED ME WRONGGGG

ezyang committed:
Update on "If the input is contiguous, short-circuit infer_size_dv in reshape"


The main improvement is that this avoids guards from infer_size_dv,
although this also counts as a minor perf improvement too.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

albanD reviewed (approved):
Sounds good!

laithsakka commented:
one recent problem is that unbacked hit this case, and when that happens we loose the option to copy as a default last resort behaviour.
